### ISSUE TITLE
False positive, `reportPossiblyUnboundVariable`

### ISSUE THREAD

**Issue by savchenko** [microsoft/pyright#8441]
**Describe the bug**
pyright reports variable as possibly unbound while there is no code path that might leave it as such. Potentially related to https://github.com/microsoft/pyright/issues/7384

**Code or Screenshots**
Here, `not_unbound` is reported as `reportPossiblyUnboundVariable`. The `elif` statement is intentional.

```python
import random
from typing import Literal

class Foo:
    @classmethod
    def bar(cls, literal_var: Literal["foo", "bar"]):
        r1 = random.choice([True, False])
        r2 = random.choice([True, False])
        if r1:
            if literal_var is "foo":
                not_unbound = 42
            elif literal_var is "bar":
                not_unbound = 24
        else:
            not_unbound = 1 if literal_var == "foo" else 0
        if r2:
            return not_unbound
```

**VS Code extension or command-line**
nVim v0.10.0, in-built LSP calling pyright v1.1.370 as `pyright-langserver --stdio`

**Comment by erictraut** [comment]
Pyright is working as intended here, so I don't consider this a bug. There's a potential enhancement request here, so I'll consider the issue in that light.

To see what's going on here, I'm going to simplify the example.

```python
def func(var: Literal["a", "b"]):
    if var is "a":
        reveal_type(var)
    else:
        reveal_type(var)
```

You can see here that the type of `var` is not narrowed in either the "if" or the "else" branch. That's because you're using a conditional expression that is not supported as a type guard in pyright. For a complete list of supported type guards, refer to [this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=type-guards).

If you change your code to use `==` rather than `is` in this case, it will work as you expect.

For comparison, mypy also does not support the `x is L` type guard pattern. 

In general, it's not safe to use `is` when comparing two literal values. It will work most of the time, but it will fail with some strings in some versions of Python. It's definitely not a recommend practice. You should use `==` and `!=` if you want your code to be safe. See [this stackoverflow post](https://stackoverflow.com/questions/1504717/why-does-comparing-strings-using-either-or-is-sometimes-produce-a-differe) for additional details.

For this reason, I don't think it makes sense for type checkers to support the `x is L` type guard pattern. For that reason, I'm going to reject this enhancement request.
